File: Dsa-day112.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> processQueries(int c, vector<vector<int>>& connections, vector<vector<int>>& queries) {
        vector<bool> online(c + 1, true);
        vector<vector<int>> adj(c + 1);
        for (auto &e : connections) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }

        vector<priority_queue<int, vector<int>, greater<int>>> substations;
        unordered_map<int, int> station_lookup;
        function<void(int, int)> dfs = [&](int node, int idx) {
            if (station_lookup.count(node)) return;
            station_lookup[node] = idx;
            substations[idx].push(node);
            for (int nei : adj[node]) dfs(nei, idx);
        };

        for (int i = 1; i <= c; i++) {
            if (!station_lookup.count(i)) {
                substations.push_back({});
                dfs(i, (int)substations.size() - 1);
            }
        }

        vector<int> ans;
        for (auto &q : queries) {
            int t = q[0], x = q[1];
            int sid = station_lookup[x];
            auto &pq = substations[sid];
            if (t == 2) {
                online[x] = false;
                while (!pq.empty() && !online[pq.top()]) pq.pop();
            } else {
                if (online[x]) ans.push_back(x);
                else {
                    while (!pq.empty() && !online[pq.top()]) pq.pop();
                    ans.push_back(pq.empty() ? -1 : pq.top());
                }
            }
        }
        return ans;
    }
};

File: Dsa-day112.py
from collections import defaultdict
from heapq import heappush, heappop

class Solution:
    def processQueries(self, c: int, connections: List[List[int]], queries: List[List[int]]) -> List[int]:
        online = [True] * (c + 1)
        adj = defaultdict(list)
        for u, v in connections:
            adj[u].append(v)
            adj[v].append(u)

        substations = []
        station_lookup = {}

        def dfs(node, idx):
            if node in station_lookup:
                return
            station_lookup[node] = idx
            heappush(substations[idx], node)
            for nei in adj[node]:
                dfs(nei, idx)

        for i in range(1, c + 1):
            if i not in station_lookup:
                substations.append([])
                dfs(i, len(substations) - 1)

        ans = []
        for t, x in queries:
            sid = station_lookup[x]
            heap = substations[sid]
            if t == 2:
                online[x] = False
                while heap and not online[heap[0]]:
                    heappop(heap)
            else:
                if online[x]:
                    ans.append(x)
                else:
                    while heap and not online[heap[0]]:
                        heappop(heap)
                    ans.append(heap[0] if heap else -1)
        return ans
